File: src/main/java/com/ottamotta/yunar/PonyService.java
package com.ottamotta.yunar;

import com.ottamotta.yunar.exceptions.CantReturnPonyException;
import com.ottamotta.yunar.exceptions.NoPonyAvailableException;
import com.ottamotta.yunar.exceptions.PonyDoesntExistsException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Isolation;
import org.springframework.transaction.annotation.Transactional;

@Service
public class PonyService {

    private AppRepository repository;

    @Autowired
    public PonyService(AppRepository repository) {
        this.repository = repository;
    }

    public Pony getStatus(String id) {
        return repository.findById(id).orElseThrow(PonyDoesntExistsException::new);
    }

    @Transactional(isolation = Isolation.REPEATABLE_READ)
    public Pony bookPony(String id) {

        Pony pony = repository.findById(id).orElseThrow(PonyDoesntExistsException::new);

        if (pony.isAvailable() && pony.isRecharged()) {
            pony.setAvailable(false);
            repository.save(pony);
            return pony;
        }

        throw new NoPonyAvailableException();
    }

    @Transactional(isolation = Isolation.REPEATABLE_READ)
    public Pony returnPony(String id) {
        Pony pony = repository.findById(id).orElseThrow(PonyDoesntExistsException::new);

        if (pony.isAvailable()) {
            throw new CantReturnPonyException();
        }

        pony.setAvailable(true);
        pony.setReturnedTimestamp(System.currentTimeMillis());
        return pony;
    }
}

File: src/main/kotlin/com/ottamotta/yunar/PonyService.kt
package com.ottamotta.yunar

import com.ottamotta.yunar.exceptions.CantReturnPonyException
import com.ottamotta.yunar.exceptions.NoPonyAvailableException
import com.ottamotta.yunar.exceptions.PonyDoesntExistsException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Isolation
import org.springframework.transaction.annotation.Transactional

@Service
class PonyService @Autowired constructor(private val repository: AppRepository) {

    fun getStatus(id: String): Pony {
        return repository.findById(id).orElseThrow { PonyDoesntExistsException() }
    }

    @Transactional(isolation = Isolation.REPEATABLE_READ)
    fun bookPony(id: String): Pony {

        val pony = repository.findById(id).orElseThrow { PonyDoesntExistsException() }

        if (pony.isAvailable && pony.isRecharged()) {
            return repository.save(pony.copy(isAvailable = false))
        }

        throw NoPonyAvailableException()
    }

    @Transactional(isolation = Isolation.REPEATABLE_READ)
    fun returnPony(id: String): Pony {
        val pony = repository.findById(id).orElseThrow { PonyDoesntExistsException() }

        if (pony.isAvailable) {
            throw CantReturnPonyException()
        }

        return repository.save(pony.copy(
                isAvailable = true,
                returnedTimestamp = System.currentTimeMillis()))

    }

    fun resetPony(name: String) {
        repository.findById(name)
                .map{it.copy(isAvailable = true, returnedTimestamp = -1) }
                .ifPresent{ repository.save(it) }
    }
}
